here-naksha-app-service/src/main/java/com/here/naksha/app/service/http/tasks/WriteFeatureApiTask.java

```diff
@@ -228,6 +234,200 @@ protected void init() {}
+    featuresFromRequest.add(featureFromRequest);
+    return attemptFeaturesPatching(spaceId, featuresFromRequest, HttpResponseType.FEATURE, addTags, removeTags, 0);
+  }
+
+  private XyzResponse attemptFeaturesPatching(
+      @NotNull String spaceId,
+      @NotNull List<XyzFeature> featuresFromRequest,
+      @NotNull HttpResponseType responseType,
+      @Nullable List<String> addTags,
+      @Nullable List<String> removeTags,
+      int retry) {
+    // Patched feature list is to ensure the order of input features is retained
+    final List<XyzFeature> patchedFeatures;
+    List<XyzFeature> featuresToPatchFromStorage = new ArrayList<>();
+    final List<String> featureIds = new ArrayList<>();
+    for (XyzFeature feature : featuresFromRequest) {
+      if (feature.getId() != null) {
+        featureIds.add(feature.getId());
+      }
+    }
+    // Extract the version of features in storage
+    final ReadFeatures rdRequest = RequestHelper.readFeaturesByIdsRequest(spaceId, featureIds);
+    try (Result result = executeReadRequestFromSpaceStorage(rdRequest)) {
+      if (result == null) {
+        return returnError(
+            XyzError.EXCEPTION,
+            "Unexpected null result while reading features from storage",
+            "Unexpected null result while reading features from storage: {}",
+            featureIds);
+      } else if (result instanceof ErrorResult er) {
+        // In case of error, convert result to ErrorResponse
+        return returnError(
+            er.reason, er.message, "Received error result while reading features in storage: {}", er);
+      }
+      try {
+        featuresToPatchFromStorage = readFeaturesFromResult(result, XyzFeature.class, 0, DEF_FEATURE_LIMIT);
+      } catch (NoCursor | NoSuchElementException emptyException) {
+        if (responseType.equals(HttpResponseType.FEATURE)) {
+          // If this is patching only 1 feature (PATCH by ID), return not found
+          return returnError(
+              XyzError.NOT_FOUND,
+              "Feature does not exist.",
+              "Unexpected null result while reading current versions in storage of targeted features for PATCH. The feature does not exist.");
+        } else if (!responseType.equals(HttpResponseType.FEATURE_COLLECTION)) {
+          // This function was then misused somewhere. FIND AND FIX IT!!
+          return returnError(
+              XyzError.EXCEPTION,
+              "Internal server error.",
+              "Unsupported HttpResponseType was called: {}",
+              responseType);
+        }
+        // Else none of the features exists in storage, will create them later
+      }
+      // Attempt patching, keeping the order of the features from the request
+      patchedFeatures = performInMemoryPatching(featuresFromRequest, featuresToPatchFromStorage);
+      // As applicable, modify features based on parameters supplied
+      for (XyzFeature featureFromRequest : patchedFeatures) {
```

Comment:
I see this as redundant iteration. Inside `performInMemoryPatching` we are already doing couple of iterations to identify which feature needs to be finally considered. You can add this `tags` logic there itself as part of same loop. There are two minor benefits:
1. we don't have to iterate again
2. all in-memory modifications to feature is happening in one place